lib/handler/mruby.c

```diff
@@ -209,6 +209,47 @@ static h2o_iovec_t convert_header_name_to_env(h2o_mem_pool_t *pool, const char *
 #undef KEY_PREFIX_LEN
 }
 
+mrb_value send_early_hints_proc(mrb_state *mrb, mrb_value self)
+{
+    mrb_value headers;
+    mrb_get_args(mrb, "H", &headers);
+
+    h2o_mruby_generator_t *generator = h2o_mruby_get_generator(mrb, mrb_proc_cfunc_env_get(mrb, 0));
+    if (generator == NULL)
+        return mrb_nil_value();
+
+    khiter_t k;
+    khash_t(ht) *h = mrb_hash_tbl(mrb, headers);
+    for (k = kh_begin(h); k != kh_end(h); ++k) {
+        if (!kh_exist(h, k))
+            continue;
+        mrb_value name = h2o_mruby_to_str(mrb, kh_key(h, k));
+        if (mrb->exc != NULL)
+            goto Failed;
+        mrb_value value = h2o_mruby_to_str(mrb, kh_value(h, k).v);
+        if (mrb->exc != NULL)
+            goto Failed;
+
+        char *pos = RSTRING_PTR(value), *end = pos + RSTRING_LEN(value);
+        while (pos < end) {
+            char *found = memchr(pos, '\n', end - pos);
+            if (found == NULL)
+                found = end;
+            h2o_add_header_by_str(&generator->req->pool, &generator->req->res.headers, RSTRING_PTR(name), RSTRING_LEN(name), 1, NULL,
+                                  pos, found - pos);
```

Comment:
Am I correct in assuming that we already have code that handles `\n` for final responses? If yes, is the logic same as that for the final response? Is there a reason for not sharing the code?

Also, I believe that removing `\n` might be the correct solution here, see [RFC 7230 section 3.2.4](https://tools.ietf.org/html/rfc7230#section-3.2.4).

Reply:
@kazuho Thank you for the comment. I fixed it at https://github.com/h2o/h2o/pull/1767/commits/e5f3a7b761b10248e8c151e3b8133f683fdf7c16

> Also, I believe that removing \n might be the correct solution here, see RFC 7230 section 3.2.4.

In Ruby web frameworks, `\n` is widely used to indicate the frameworks to send multiple headers with a same name at once.